common/src/main/kotlin/com/microsoft/identity/common/BrokerDiscoveryClient.kt

```diff
@@ -0,0 +1,166 @@
+import com.microsoft.identity.common.internal.broker.PackageHelper
+import com.microsoft.identity.common.internal.broker.ipc.BrokerOperationBundle
+import com.microsoft.identity.common.internal.broker.ipc.ContentProviderStrategy
+import com.microsoft.identity.common.internal.broker.ipc.IIpcStrategy
+import com.microsoft.identity.common.logging.Logger
+import kotlinx.coroutines.*
+import java.text.SimpleDateFormat
+import java.util.*
+
+class BrokerDiscoveryClient(private val brokerCandidates: Set<BrokerData>,
+                            private val getActiveBrokerFromAccountManager: () -> BrokerData?,
+                            private val ipcStrategy: IIpcStrategy,
+                            private val cache: BrokerMetadataCache,
+                            private val isPackageInstalled: (BrokerData) -> Boolean): IBrokerDiscoveryClient {
+
+    companion object{
+        val TAG = BrokerDiscoveryClient::class.simpleName
+
+
+
+        private val BROKER_CANDIDATE_SET = setOf(
+            BrokerData.MOCK_LTW,
+            BrokerData.MOCK_AUTHAPP,
+            BrokerData.MOCK_CP,
+        )
+
+        const val ACTIVE_BROKER_PACKAGE_NAME_BUNDLE_KEY = "ACTIVE_BROKER_PACKAGE_NAME_BUNDLE_KEY"
+        const val ACTIVE_BROKER_SIGNATURE_HASH_BUNDLE_KEY = "ACTIVE_BROKER_SIGNATURE_HASH_BUNDLE_KEY"
+        const val BENCHMARK_STRING_BUNDLE_KEY = "BENCHMARK_STRING_BUNDLE_KEY"
+        const val ERROR_BUNDLE_KEY = "ERROR_BUNDLE_KEY"
+        const val FORCE_DISCOVERY_BUNDLE_KEY = "FORCE_DISCOVERY_BUNDLE_KEY"
+
+        // TODO: add tests on this.
+        fun getActiveBrokerFromAccountManagerFunc(context: Context) : () -> BrokerData? {
+            return {
+                val activeBroker = BrokerValidator(context).currentActiveBrokerPackageName
+                activeBroker?.let {
+                    BrokerData(activeBroker,
+                        PackageHelper(context.packageManager).getCurrentSignatureForPackage(activeBroker))
+                }
+            }
+        }
+
+        @SuppressLint("SimpleDateFormat")
+        fun getCurrentTime(): String {
+            val date = Date(System.currentTimeMillis())
+            val formatter = SimpleDateFormat("hh:mm:ss.SSS");
+            return formatter.format(date)
+        }
+
+        private fun queryFromBroker(brokerCandidates: Set<BrokerData>,
+                                    ipcStrategy: IIpcStrategy,
+                                    isPackageInstalled: (BrokerData) -> Boolean,
+                                    forceTrigger: Boolean): BrokerData? {
+            val methodTag = "$TAG:queryFromBroker"
+
+            val bundle = Bundle()
+            bundle.putBoolean(FORCE_DISCOVERY_BUNDLE_KEY, forceTrigger)
+
+            // ping all candidates to frontload bootstrapping!
+            val installedCandidates = brokerCandidates.filter(isPackageInstalled)
+
+            val list = mutableListOf<Deferred<Unit>>()
```

Comment:
Add tests where broker discovery on multiple broker happens multiple time.
How are we making sure the result is NOT corrupted?

(technically... they should yield the same result... but just in case... maybe use timestamp when the whole flow is initiated?)

Reply:
apart from that... maybe we still don't need to parallelize this .